tools/cli/pkg/command/taskrun.go

```diff
@@ -44,29 +91,272 @@ For each subprocess, the following Runtime-specific data are passed as environme
+	cobraCmd.Flags().BoolVar(&cmd.keepKubeconfigs, "keep-kubeconfig", false, "Option that allows you to keep downloaded kubeconfig files after execution for caching purposes.")
+	cobraCmd.Flags().BoolVar(&cmd.noKubeconfig, "no-kubeconfig", false, "Option that turns off the downloading and exposure of the kubeconfig file for each Runtime.")
+	cobraCmd.Flags().BoolVar(&cmd.noPrefixOutput, "no-prefix-output", false, "Option that omits the prefixing of each output line with the Runtime name. By default, all output lines are prepended for better traceability.")
 	return cobraCmd
 }
 
 // Run executes the taskrun command
-func (cmd *TaskRunCommand) Run() error {
-	fmt.Println("Not implemented yet.")
-	return nil
+func (cmd *TaskRunCommand) Run(args []string) error {
+	cmd.log = logger.New()
+	cmd.cred = CLICredentialManager(cmd.log)
+	defer cmd.cleanupTempKubeConfigDir()
+
+	operations, err := cmd.resolveOperations()
+	if err != nil {
+		return err
+	}
+
+	mgr := NewRuntimeTaskMakager(cmd, operations)
+	strategy := orchestration.NewParallelOrchestrationStrategy(mgr, cmd.log)
+	execID, err := strategy.Execute(operations, orchestration.StrategySpec{
+		Type:     orchestration.ParallelStrategy,
+		Schedule: orchestration.Immediate,
+		Parallel: orchestration.ParallelStrategySpec{Workers: cmd.parallelism},
+	})
+	if err != nil {
+		return errors.Wrap(err, "while executing task")
+	}
+	strategy.Wait(execID)
+
+	return mgr.exitStatus()
 }
 
 // Validate checks the input parameters of the taskrun command
-func (cmd *TaskRunCommand) Validate() error {
+func (cmd *TaskRunCommand) Validate(args []string) error {
+	// Validate kubeconfig-api-url global option
+	if !cmd.noKubeconfig && GlobalOpts.KubeconfigAPIURL() == "" {
+		return fmt.Errorf("missing required %s option", GlobalOpts.kubeconfigAPIURL)
+	}
+
+	// Validate gardener-kubeconfig global option
+	if GlobalOpts.GardenerKubeconfig() == "" || GlobalOpts.GardenerNamespace() == "" {
+		return fmt.Errorf("missing required %s/%s options", GlobalOpts.gardenerKubeconfig, GlobalOpts.gardenerNamespace)
+	}
+
+	// Validate target options
 	err := ValidateTransformRuntimeTargetOpts(cmd.targetInputs, cmd.targetExcludeInputs, &cmd.targets)
 	if err != nil {
 		return err
 	}
-	// TODO: check if cmd.kubeconfigDir dir exists if input was given
+
+	// Validate kubeconfig directory
+	if cmd.kubeconfigDir != "" {
+		fi, err := os.Stat(cmd.kubeconfigDir)
+		if err != nil {
+			return err
+		}
+		if !fi.IsDir() {
+			return fmt.Errorf("%s: not a directory", cmd.kubeconfigDir)
+		}
+	} else if !cmd.noKubeconfig {
+		cmd.kubeconfigDir, err = ioutil.TempDir("", "kubeconfig-")
+		if err != nil {
+			return errors.Wrap(err, "while creating temporary kubeconfig directory")
+		}
+		cmd.kubeconfingDirTemp = true
+	}
+
+	// Validate task command
+	if _, err := exec.LookPath(args[0]); err != nil {
+		return err
+	}
+	cmd.taskCommand = exec.CommandContext(cmd.cobraCmd.Context(), args[0], args[1:]...)
```

Comment:
Unneccessarry expansion
```go
cmd.taskCommand = exec.CommandContext(cmd.cobraCmd.Context(), args...)
```

Reply:
It's not unnecessary. The called function's signature is `func exec.CommandContext(ctx context.Context, name string, arg ...string) *exec.Cmd`

So caller must assign the second argument, and may pass subsequent arguments.